Cython/Compiler/ExprNodes.py

```diff
@@ -10184,6 +10187,93 @@ def calculate_result_code(self):
+    #  C++ typeid operator applied to a type or variable
+    #
+    #  operand       ExprNode
+    #  arg_type      ExprNode
+    #  is_variable   boolean
+    #  mangle_cname  string
+
+    type = PyrexTypes.error_type
+
+    subexprs = ['operand']
+
+    arg_type = None
+    is_variable = None
+    mangle_cname = None
+
+    def get_type_info_type(self, env):
+        if env.is_module_scope:
+            env_module = env
+        else:
+            env_module = env.outer_scope
+        for module in env_module.cimported_modules:
+            if module.qualified_name == 'libcpp.typeinfo':
+                type_info = module.lookup('type_info')
+                type_info = PyrexTypes.c_ref_type(PyrexTypes.c_const_type(type_info.type))
+                return type_info
+        return None
+
+    def analyse_types(self, env):
+        type_info = self.get_type_info_type(env)
+        if not type_info:
+            self.error("The 'libcpp.typeinfo' module must be cimported to use the typeid() operator")
+            return self
+        self.type = type_info
+        as_type = self.operand.analyse_as_type(env)
+        if as_type:
+            self.arg_type = as_type
+            self.is_type = True
+        else:
+            self.arg_type = self.operand.analyse_types(env)
+            self.is_type = False
+            if self.arg_type.type.is_pyobject:
+                self.error("Cannot use typeid on a Python object")
+                return self
+            elif self.arg_type.type.is_void:
+                self.error("Cannot use typeid on void")
+                return self
+            elif not self.arg_type.type.is_complete():
+                self.error("Cannot use typeid on incomplete type '%s'" % self.arg_type.type)
+                return self
+        if env.is_module_scope:
+            env_module = env
+        else:
+            env_module = env.outer_scope
+        env_module.typeid_variables += 1
+        self.mangle_cname = "%s_typeid_%s" % (
+            env_module.module_cname, env_module.typeid_variables)
+        env.use_utility_code(UtilityCode.load_cached("CppExceptionConversion", "CppSupport.cpp"))
+        return self
+
+    def error(self, mess):
+        error(self.pos, mess)
+        self.type = PyrexTypes.error_type
+        self.result_code = "<error>"
+
+    def check_const(self):
+        return True
+
+    def calculate_result_code(self):
+        return "(*%s)" % self.mangle_cname
+
+    def generate_result_code(self, code):
+        if self.is_type:
+            if self.arg_type.is_extension_type:
+                # the size of the pointer is boring
+                # we want the size of the actual struct
+                arg_code = self.arg_type.declaration_code("", deref=1)
+            else:
+                arg_code = self.arg_type.empty_declaration_code()
+        else:
+            arg_code = self.arg_type.result()
+        code.putln("const std::type_info *%s;" % self.mangle_cname)
```

Comment:
I did not use `self.result()` there because I needed to dereference the temporary variable in `calculate_result_code`; I needed to make the temp var be of the type `const std::type_info *` so I could initialize it and then assign to it in the try-catch block.

> Look for usages of allocate_temp() in this module

Thanks! I did not notice `allocate_temp()` yet, I will change it around to do that instead!